plugins/outputs/zabbix/lld.go

```diff
@@ -0,0 +1,196 @@
+		}
+
+		// Add the LLD m
+		m := metric.New(
+			lldName,
+			map[string]string{
+				"host": key.Hostname,
+			},
+			map[string]interface{}{
+				key.LLDKey: dataValuesJSON,
+			},
+			time.Now(),
+		)
+		metrics = append(metrics, m)
+	}
+
+	// Send empty LLDs for the LLDs that were sent the last time but not this time.
+	for key := range zl.previousReceivedData {
+		emptyDataValuesJSON, err := json.Marshal(map[string]interface{}{"data": []interface{}{}})
+		if err != nil {
+			zl.log.Warnf("Marshaling to JSON empty data Zabbix format: %v", err)
+		}
+
+		// Add the LLD m
+		m := metric.New(
+			lldName,
+			map[string]string{
+				"host": key.Hostname,
+			},
+			map[string]interface{}{
+				key.LLDKey: emptyDataValuesJSON,
+			},
+			time.Now(),
+		)
+		metrics = append(metrics, m)
+	}
+
+	// Move current data to previousReceivedData
+	zl.previousReceivedData = zl.receivedData
+
+	// Reset receivedData and cache to receive data from the new period
+	zl.receivedData = make(map[hostLLD]lldTags, len(zl.receivedData))
+	zl.cache = make(map[uint64]bool, len(zl.cache))
+
+	// Clear previousReceivedData if enough time has passed
+	if time.Since(zl.lastClear) > time.Duration(zl.clearInterval) {
+		zl.previousReceivedData = make(map[hostLLD]lldTags, len(zl.previousReceivedData))
+		zl.lastClear = time.Now()
+	}
+
+	return metrics
+}
+
+// Add parse a metric and add it to the LLD cache.
+func (zl *zabbixLLD) Add(in telegraf.Metric) {
+	hostname, err := getHostname(in.TagList())
+	if err != nil {
+		zl.log.Errorf("Error getting hostname: %v", err)
+		return
+	}
+
+	// HashID generates an id based on the measurement, tags keys and tags values
+	id := in.HashID()
+	if _, ok := zl.cache[id]; !ok {
+		// hit an uncached metric, store in the lld with metrics of the same kind
+		zl.cache[id] = true
+
+		lldKey, err := generateLLDKey(in.Name(), in.TagList())
+		if err != nil {
+			// Metric with no tags (or just the "host" tag), or no measurement name are ignored.
+			return
+		}
+
+		hostLLDKey := hostLLD{hostname, lldKey}
+
+		hostLLD, exists := zl.receivedData[hostLLDKey]
+		if !exists {
+			hostLLD = lldTags{}
+		}
+
+		// tags without the "host" tag
+		tags := make(map[string]string, len(in.TagList()))
+		for _, tag := range in.TagList() {
+			if tag.Key == "host" {
+				continue
+			}
+			tags[tag.Key] = tag.Value
+		}
+
+		hostLLD = append(hostLLD, tags)
+		zl.receivedData[hostLLDKey] = hostLLD
+	}
+}
+
+// compareAndDelete returns true if the key has been already sent with the same tags.
+// It compares the values of "tags" with the ones stored in p.previousReceivedData
+// for the key specified.
+// Also delete the key in p.previousReceivedData, to know which keys remains.
+func (zl *zabbixLLD) compareAndDelete(key hostLLD, tags lldTags) (equal bool) {
+	previousTags, exists := zl.previousReceivedData[key]
+	if !exists {
+		return false
+	}
+	delete(zl.previousReceivedData, key)
+
+	return elementsMatch(tags, previousTags)
```

Comment:
Please get rid of `elementsMatch`! I would implement a hash function for `lldTags` (similar to [this code](https://goplay.space/#ykiyPpy8AtD)) and then compare the hashes. This should be much faster and you might be able to cache the previous hashes e.g. in `previousReceivedData`...

Reply:
Fixed in d815ad6e2